kvserver: add observability (metrics & logging) for decommissioning nudger

dodeca12 opened the pull request:
As titled - currently there's no way to know if the decommissioning nudger works as intended or not. PR adds observability to ensure the nudger's working as intended.

>Note: `DecommissioningNudgerEqneueFailure` metric is rarely incremented, as the `maybeEnqueueProblemRange` function is expected to be single-threaded. Additionally, `DecommissioningNudgerEnqueueEnqueued` is worded that way and not `DecommissioningNudgerEnqueuedSucceess` as it is inaccurate to say "success" implies a successful enqueue since `.AddAsync` only schedules the replica for processing but does not guarantee that the operation will complete successfully or that the replica will actually be processed by the queue. These peculiarities suggest there's merit in revisiting and reworking the metrics (and perhaps logging).

#### Testing
```sh
$ ./dev test pkg/kv/kvserver -f TestReplicateQueueDecommissionScannerDisabled -v --show-logs

# also good to run the `--stress` and `--race` flags to ensure no flakiness and resolve CI headaches:
$ ./dev test pkg/kv/kvserver -f TestReplicateQueueDecommissionScannerDisabled -v --show-logs --stress --race
```

Fixes https://github.com/cockroachdb/cockroach/issues/148090
Epic: None

iskettaneh reviewed (commented):
look good :fire: .. just added a few nits and some clarifying questions.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-:-OXet-SN3IFU5guteXnp:b14mtoa)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvserver/replica.go` line 2943 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXepkEXD7-jpQDgQ6hj:-OXepkEYCNfmyfywVGW1:buq0564) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replica.go#L2943)):*
> ```Go
> 		if !isLeaseholder {
> 			if problemRangeLogLimiter.ShouldLog() {
> 				log.KvDistribution.VInfof(ctx, 1, "Not enqueuing replica %s because it is not the leaseholder", r.Desc())
> ```

@arulajmani I know you had a comment about logging every minute. Did you mean that all logs here should be logged every minute? Or only when the log line that says not enqueuing because setting is disabled?
___
*[`pkg/kv/kvserver/replica.go` line 2974 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXeqZLZ7bnaewpMZxkb:-OXeqZLZ7bnaewpMZxkc:b-sbo1bc) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replica.go#L2974)):*
> ```Go
> 	if !r.lastProblemRangeReplicateEnqueueTime.CompareAndSwap(lastTime, now) {
> 		if problemRangeLogLimiter.ShouldLog() {
> 			log.KvDistribution.VInfof(ctx, 1, "Not enqueuing replica %s because the last time the replica was enqueued is less than the interval ago", r.Desc())
> ```

nit: we generally start our logs in a lower case character.
___
*[`pkg/kv/kvserver/replica.go` line 2976 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXeq9pHAKlHz6nNIvEs:-OXeq9pHAKlHz6nNIvEt:b-10jiwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replica.go#L2976)):*
> ```Go
> 			log.KvDistribution.VInfof(ctx, 1, "Not enqueuing replica %s because the last time the replica was enqueued is less than the interval ago", r.Desc())
> 		}
> 		// Note: This metric is rarely incremented, as maybeEnqueueProblemRange is expected to be single-threaded.
> ```

nit: the general guideline is to wrap comments at 80 characters: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Line-Length
___
*[`pkg/kv/kvserver/replica.go` line 2976 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXerIX35cW1vDXwC9fb:-OXerIX35cW1vDXwC9fc:b-7kpvdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replica.go#L2976)):*
> ```Go
> 			log.KvDistribution.VInfof(ctx, 1, "Not enqueuing replica %s because the last time the replica was enqueued is less than the interval ago", r.Desc())
> 		}
> 		// Note: This metric is rarely incremented, as maybeEnqueueProblemRange is expected to be single-threaded.
> ```

Do we know of any case where this could fail?
___
*[`pkg/kv/kvserver/replicate_queue_test.go` line 2485 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXesoSw7PiKWFKk3Nt0:-OXesoSw7PiKWFKk3Nt1:bp4m1ki) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replicate_queue_test.go#L2485)):*
> ```Go
> 	// Wait for the enqueue logic to trigger and validate metrics were updated.
> 	testutils.SucceedsSoon(t, func() error {
> 		afterAttempts := getDecommissioningNudgerMetricValue(t, tc, "attempts")
> ```

Shall we to assert on the other metrics as well?
___
*[`pkg/kv/kvserver/replicate_queue_test.go` line 2510 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXerj6ZDeN8FH_l8z9H:-OXerj6ZDeN8FH_l8z9I:bhq1qx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replicate_queue_test.go#L2510)):*
> ```Go
> // getDecommissioningNudgerMetricValue retrieves the value of a specific decommissioning
> // nudger metric from the test cluster. It aggregates the values across all stores.
> func getDecommissioningNudgerMetricValue(
> ```

nit: Since this function is only used inside one test, you could make it anonymous function.
For example, inside the test, you could say:

getDecomMetricValue := func(metricType string) {
 ... put the logic here ...
}

and then later in the test you could just call it: getDecomMetricValue("attempts")


<!-- Sent from Reviewable.io -->

dodeca12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @miraradeva)

___
*[`pkg/kv/kvserver/replica.go` line 2976 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXerIX35cW1vDXwC9fb:-OXjD5CJ7X8uB1Stjapf:b-6loeaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replica.go#L2976)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Do we know of any case where this could fail?
</blockquote></details>

Assuming the code doesn't change, this can fail if there are concurrent calls to `maybeEnqueueProblemRange` for the same replica. But I think it's safe to assume that the function is only expected to run when it's single-threaded 
___
*[`pkg/kv/kvserver/replicate_queue_test.go` line 2485 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXesoSw7PiKWFKk3Nt0:-OXjBt3GD4A4cZLgp0GY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replicate_queue_test.go#L2485)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Shall we to assert on the other metrics as well?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Reviewed 3 of 4 files at r3, 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @miraradeva)

___
*[`pkg/kv/kvserver/replica.go` line 2943 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXepkEXD7-jpQDgQ6hj:-OXjJa7v9P5DGYGCCMvP:b-hosxfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replica.go#L2943)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

@arulajmani I know you had a comment about logging every minute. Did you mean that all logs here should be logged every minute? Or only when the log line that says not enqueuing because setting is disabled?
</blockquote></details>

I meant just the cluster setting disabled log line. 
___
*[`pkg/kv/kvserver/replica.go` line 2941 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjJuZQ7_uVShvVbG2N:-OXjJuZQ7_uVShvVbG2O:b8op5ar) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2941)):*
> ```Go
> 		// The replicate queue will not process the replica without a valid lease.
> 		// Nothing to do.
> 		if !isLeaseholder {
> ```

nit: let's reduce some nesting here, even once we buy back some real estate by getting rid of the problemRangeLogLimiter.
___
*[`pkg/kv/kvserver/replica.go` line 2976 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjK2fk56J7agAxPMcI:-OXjK2fk56J7agAxPMcJ:bwek33b) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2976)):*
> ```Go
> 			log.KvDistribution.VInfof(ctx, 1, "not enqueuing replica %s because the last time the replica was enqueued is less than the interval ago", r.Desc())
> 		}
> 		// Note: This metric is rarely incremented, as maybeEnqueueProblemRange is expected to be single-threaded.
> ```

nit: Is this comment wrapped to 80 chars?
___
*[`pkg/kv/kvserver/replica.go` line 2979 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjKCTg2H6cscOxTV93:-OXjKCTg2H6cscOxTV94:bt5x0w3) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2979)):*
> ```Go
> 		// It is kept for completeness, but may not be useful in practice.
> 		if r.store.metrics.DecommissioningNudgerEnqueueFailure != nil {
> 			r.store.metrics.DecommissioningNudgerEnqueueFailure.Inc(1)
> ```

Did you have a chance to speak with @wenyihu6 about this? She correctly pointed out that keeping this metric as is might be misleading, as IIUC, the real failure to enqueue something from the decommissioning nudger will happen below `AddAsync`.
___
*[`pkg/kv/kvserver/replica.go` line 2986 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjKU094picUKbO6sQP:-OXjKU094picUKbO6sQQ:b4j5pmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2986)):*
> ```Go
> 	// (other logs have a verbosity of 1 which).
> 	log.KvDistribution.Infof(ctx, "decommissioning nudger enqueuing replica %s with priority %f", r.Desc(), allocatorimpl.AllocatorReplaceDecommissioningVoter.Priority())
> 	if r.store.metrics.DecommissioningNudgerEnqueueEnqueued != nil {
> ```

Why does this need a nil check?


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
Looks good. Thanks for working on this - will be handy for escalations involving decommission stall!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-:-OXxqBoU27FphJuKltEi:b5hnd4w)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @dodeca12, @iskettaneh, and @miraradeva)

___
*[`pkg/kv/kvserver/replica.go` line 2979 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjKCTg2H6cscOxTV93:-OXxgeGP1nj_LZN7xbQc:b-q4zb8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2979)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Did you have a chance to speak with @wenyihu6 about this? She correctly pointed out that keeping this metric as is might be misleading, as IIUC, the real failure to enqueue something from the decommissioning nudger will happen below `AddAsync`.
</blockquote></details>

This case doesn't seem that interesting since it feels pretty rare. The only situation we might hit is if https://github.com/cockroachdb/cockroach/blob/a2e2f92c764f2bb5b5cf556b266df753bdb06866/pkg/kv/kvserver/store_gossip.go#L194,  https://github.com/cockroachdb/cockroach/blob/31db746035d2943f7034818f6dc309971718a2bc/pkg/kv/kvserver/store.go#L3751, or https://github.com/cockroachdb/cockroach/blob/f17cf577363c6a128ce5bf922da5691483ff61e6/pkg/kv/kvserver/replica_init.go#L190 are called concurrently. But https://github.com/cockroachdb/cockroach/blob/a2e2f92c764f2bb5b5cf556b266df753bdb06866/pkg/kv/kvserver/store_gossip.go#L194 is only called once during store initialization and https://github.com/cockroachdb/cockroach/blob/f17cf577363c6a128ce5bf922da5691483ff61e6/pkg/kv/kvserver/replica_init.go#L190 is for replica initialization. I also don't feel like this counts as a failure.

It feels more interesting if we called into `AddAsync` but didn't end up calling addLocked on the queue https://github.com/cockroachdb/cockroach/blob/32f7bad014fe3d72f54ab9c2296d34ff47cfbecc/pkg/kv/kvserver/queue.go#L767. Currently, the code in place doesn't make this easy since it's async, we would need some channel to pass the result back. If this is too large of a change, I would prefer moving the failure metrics to track when we are not the leaseholder or the lease is invalid (if we are here, range counter is true, https://github.com/cockroachdb/cockroach/blob/e610dd89a058942a5e54d90dd77621d66cb19b65/pkg/kv/kvserver/replica_metrics.go#L271, so I would expect it to be true most of the time.) Wdyt?
___
*[`pkg/kv/kvserver/replica.go` line 2986 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjKU094picUKbO6sQP:-OXxhtFX75MiEsNyD0u_:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2986)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why does this need a nil check?
</blockquote></details>

+1
___
*[`pkg/kv/kvserver/replica.go` line 2931 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXxdzUJ9EYGhY9hFGwx:-OXxdzUJ9EYGhY9hFGwy:behahel) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15f37d3f4daddf89db46fa1ca281a7a2cab53ac/pkg/kv/kvserver/replica.go#L2931)):*
> ```Go
> 	// Increment attempts metric for every call.
> 	if r.store.metrics.DecommissioningNudgerEnqueueAttempts != nil {
> 		r.store.metrics.DecommissioningNudgerEnqueueAttempts.Inc(1)
> ```

This feels a bit repetitive as https://github.com/cockroachdb/cockroach/blob/31db746035d2943f7034818f6dc309971718a2bc/pkg/kv/kvserver/store.go#L3492, https://github.com/cockroachdb/cockroach/blob/173547ec7c8135c50a3f811144f9d3b9bd8195f5/pkg/kv/kvserver/metrics.go#L167. I would have expected maybeEnqueueProblemRange to be called for every decommissioning range.
___
*[`pkg/kv/kvserver/replica.go` line 2952 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXxeZLb1GmZZFDjmVYJ:-OXxeZLb1GmZZFDjmVYK:b65m02f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15f37d3f4daddf89db46fa1ca281a7a2cab53ac/pkg/kv/kvserver/replica.go#L2952)):*
> ```Go
> 	if interval == 0 {
> 		// The setting is disabled.
> 		if problemRangeLogLimiter.ShouldLog() {
> ```

I’d prefer if we don't log this (especially every minute). An interval of 0 can already be seen from the cluster setting in the debug zip.


<!-- Sent from Reviewable.io -->

dodeca12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, @miraradeva, and @wenyihu6)

___
*[`pkg/kv/kvserver/replica.go` line 2943 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXepkEXD7-jpQDgQ6hj:-OXjX_Xq18hQTyt195Le:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/678ec2385d610003c6016090caedaf8eb5f319e7/pkg/kv/kvserver/replica.go#L2943)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I meant just the cluster setting disabled log line.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica.go` line 2979 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjKCTg2H6cscOxTV93:-OXjZ9Nj91ohuXnu3WMy:b-ninl39) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2979)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

This case doesn't seem that interesting since it feels pretty rare. The only situation we might hit is if https://github.com/cockroachdb/cockroach/blob/a2e2f92c764f2bb5b5cf556b266df753bdb06866/pkg/kv/kvserver/store_gossip.go#L194,  https://github.com/cockroachdb/cockroach/blob/31db746035d2943f7034818f6dc309971718a2bc/pkg/kv/kvserver/store.go#L3751, or https://github.com/cockroachdb/cockroach/blob/f17cf577363c6a128ce5bf922da5691483ff61e6/pkg/kv/kvserver/replica_init.go#L190 are called concurrently. But https://github.com/cockroachdb/cockroach/blob/a2e2f92c764f2bb5b5cf556b266df753bdb06866/pkg/kv/kvserver/store_gossip.go#L194 is only called once during store initialization and https://github.com/cockroachdb/cockroach/blob/f17cf577363c6a128ce5bf922da5691483ff61e6/pkg/kv/kvserver/replica_init.go#L190 is for replica initialization. I also don't feel like this counts as a failure.

It feels more interesting if we called into `AddAsync` but didn't end up calling addLocked on the queue https://github.com/cockroachdb/cockroach/blob/32f7bad014fe3d72f54ab9c2296d34ff47cfbecc/pkg/kv/kvserver/queue.go#L767. Currently, the code in place doesn't make this easy since it's async, we would need some channel to pass the result back. If this is too large of a change, I would prefer moving the failure metrics to track when we are not the leaseholder or the lease is invalid (if we are here, range counter is true, https://github.com/cockroachdb/cockroach/blob/e610dd89a058942a5e54d90dd77621d66cb19b65/pkg/kv/kvserver/replica_metrics.go#L271, so I would expect it to be true most of the time.) Wdyt?
</blockquote></details>

Moved it down closer to the source of the actual failures in the next commit
___
*[`pkg/kv/kvserver/replica.go` line 2986 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXjKU094picUKbO6sQP:-OY1RT521fuBtP2MfMVm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f3a3adadb934f742a8e5bc960dcd3cf198ae12/pkg/kv/kvserver/replica.go#L2986)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica.go` line 2931 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXxdzUJ9EYGhY9hFGwx:-OXyJexhCaiz43KIqnng:b7cxqrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15f37d3f4daddf89db46fa1ca281a7a2cab53ac/pkg/kv/kvserver/replica.go#L2931)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

This feels a bit repetitive as https://github.com/cockroachdb/cockroach/blob/31db746035d2943f7034818f6dc309971718a2bc/pkg/kv/kvserver/store.go#L3492, https://github.com/cockroachdb/cockroach/blob/173547ec7c8135c50a3f811144f9d3b9bd8195f5/pkg/kv/kvserver/metrics.go#L167. I would have expected maybeEnqueueProblemRange to be called for every decommissioning range.
</blockquote></details>

Sure I'll remove it
___
*[`pkg/kv/kvserver/replica.go` line 2952 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OXxeZLb1GmZZFDjmVYJ:-OXyJjC9D8DVDKA-eICu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e15f37d3f4daddf89db46fa1ca281a7a2cab53ac/pkg/kv/kvserver/replica.go#L2952)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I’d prefer if we don't log this (especially every minute). An interval of 0 can already be seen from the cluster setting in the debug zip.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @miraradeva)

___
*[`pkg/kv/kvserver/queue.go` line 711 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY1Y3C-4wantYgxvHO0:-OY1Y3C-4wantYgxvHO1:b-76vpc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7b25566c305d3843be7b58ed8ff7ef384bd565/pkg/kv/kvserver/queue.go#L711)):*
> ```Go
> 	// `allocatorimpl.AllocatorReplaceDecommissioningVoter.Priority()`.
> 	return bq.name == "replicate" &&
> 		priority == allocatorimpl.AllocatorReplaceDecommissioningVoter.Priority()
> ```

Lets discuss during our 1:1 today. `allocatorimpl.AllocatorReplaceDecommissioningVoter.Priority()` is not only enqueued by the nudger. It can also come from the replicate queue directly. 
___
*[`pkg/kv/kvserver/queue.go` line 738 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY1YMuMCsllctKYqrqR:-OY1YMuMCsllctKYqrqS:b50t5ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7b25566c305d3843be7b58ed8ff7ef384bd565/pkg/kv/kvserver/queue.go#L738)):*
> ```Go
> 		// Track decommissioning nudger failures.
> 		if bq.isDecommissioningNudgerEnqueue(ctx, desc, priority) {
> 			bq.trackDecommissioningNudgerFailure(ctx, desc, errors.New("replica not initialized"))
> ```

nit: I’d prefer just bumping the metrics directly here instead of using a helper function


<!-- Sent from Reviewable.io -->

dodeca12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, @miraradeva, and @wenyihu6)

___
*[`pkg/kv/kvserver/queue.go` line 738 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY1YMuMCsllctKYqrqR:-OY20O-H9c2ErljZEsrr:btt30sj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d7b25566c305d3843be7b58ed8ff7ef384bd565/pkg/kv/kvserver/queue.go#L738)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: I’d prefer just bumping the metrics directly here instead of using a helper function
</blockquote></details>

if it were just bumping metrics sure, but I'm also adding logging, so wrapping it in a helper makes the code cleaner


<!-- Sent from Reviewable.io -->

Commit:
kvserver: add logging & metrics for decommissioning nudger

Previously, the decommissioning nudger had limited observability, making it
difficult to monitor its effectiveness and diagnose issues during node
decommissioning operations.

This was inadequate because operators couldn't track how many ranges were
being enqueued for decommissioning, nor could they see when the nudger
skipped ranges due to leaseholder status or invalid leases.

To address this, this patch adds comprehensive logging and metrics:
- Logs when the decommissioning nudger enqueues replicas with priority info
- Tracks successful enqueues via DecommissioningNudgerEnqueueEnqueued metric
- Tracks skipped enqueues via
DecommissioningNudgerNotLeaseholderOrInvalidLease metric
- Adds structured logging for debugging nudger behavior
- Includes comprehensive test coverage for the new metrics

TODO:
- Figure out a better way to track decommissioning enqueue failures.
	Currently it's hard to "get as close as we can" to the source of the
	enqueue failures for logging & metrics purposes - this would require a
	better architecting of the code pathways to ensure we log and track
	failures as close as we can to where they occur

Fixes: CRDB-51396
Release note: None

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work! Preemptive approve mod the comments. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-:-OY75wGQ5Ex4mVRITA14:b-3ltpv3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @dodeca12, @iskettaneh, and @miraradeva)

___
*[`pkg/kv/kvserver/metrics.go` line 181 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY74Xn-D7-akFL_gLdJ:-OY74Xn-D7-akFL_gLdK:b3cd2cn) ([raw file](https://github.com/cockroachdb/cockroach/blob/479d65569f150873a41fdd6834d0d16ddc28357c/pkg/kv/kvserver/metrics.go#L181)):*
> ```Go
> 		Unit:         metric.Unit_COUNT,
> 		LabeledName:  "ranges.decommissioning.nudger.enqueue.enqueued",
> 		StaticLabels: metric.MakeLabelPairs(metric.LabelStatus, "enqueued"),
> ```

nit: This might be useful, but it doesn’t seem consistent with the conventions in this part of the code. What are `LabeledName` and `StaticLabels` used for here? It would be nice to keep things symmetric with `metaDecommissioningRangeCount` for now. We can add them for all if needed in the future. 

nit: While we are here, do you mind also adding a sentence here explaining - it might not end up being enqueued by the priority queue in the description. Rename `ranges.decommissioning.nudger.enqueue.enqueued` to `ranges.decommissioning.nudger.enqueue` would also be nice. 
___
*[`pkg/kv/kvserver/replica.go` line 2936 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY76GysBzdQvsLpRaiu:-OY76GysBzdQvsLpRaiv:b-442h43) ([raw file](https://github.com/cockroachdb/cockroach/blob/479d65569f150873a41fdd6834d0d16ddc28357c/pkg/kv/kvserver/replica.go#L2936)):*
> ```Go
> 		if !isLeaseholder {
> 			log.KvDistribution.VInfof(ctx, 1, "not enqueuing replica %s because it is not the leaseholder", r.Desc())
> 			r.store.metrics.DecommissioningNudgerNotLeaseholderOrInvalidLease.Inc(1)
> ```

Instead of the nesting here, could we print two booleans in one log statements. 
___
*[`pkg/kv/kvserver/replica.go` line 2962 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY73YV8CG74tMIvlUUV:-OY73YV8CG74tMIvlUUW:b-7j8x2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/479d65569f150873a41fdd6834d0d16ddc28357c/pkg/kv/kvserver/replica.go#L2962)):*
> ```Go
> 	if !r.lastProblemRangeReplicateEnqueueTime.CompareAndSwap(lastTime, now) {
> 		// This race condition is expected to be rare.
> 		log.KvDistribution.VInfof(ctx, 1, "not enqueuing replica %s because the last time the replica was enqueued is less than the interval ago", r.Desc())
> ```

nit: It's not really interval ago here - we are caught up in a race condition. Interval ago would be for `if lastTime.Add(interval).After(now)`.


<!-- Sent from Reviewable.io -->

dodeca12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151898)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @arulajmani, @iskettaneh, @miraradeva, and @wenyihu6)

___
*[`pkg/kv/kvserver/metrics.go` line 181 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY74Xn-D7-akFL_gLdJ:-OY7F31k7tKSNoEKf23X:b-xhivch) ([raw file](https://github.com/cockroachdb/cockroach/blob/7999aa14f5556fee86351c9cf336c0e306a1c5d1/pkg/kv/kvserver/metrics.go#L181)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: This might be useful, but it doesn’t seem consistent with the conventions in this part of the code. What are `LabeledName` and `StaticLabels` used for here? It would be nice to keep things symmetric with `metaDecommissioningRangeCount` for now. We can add them for all if needed in the future. 

nit: While we are here, do you mind also adding a sentence here explaining - it might not end up being enqueued by the priority queue in the description. Rename `ranges.decommissioning.nudger.enqueue.enqueued` to `ranges.decommissioning.nudger.enqueue` would also be nice.
</blockquote></details>

This was suggested by @alyshanjahani-crl - see https://github.com/cockroachdb/cockroach/pull/151678#discussion_r2277579900
___
*[`pkg/kv/kvserver/replica.go` line 2936 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/151898#-OY76GysBzdQvsLpRaiu:-OY7FYO82cB5K40Mf3I-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7999aa14f5556fee86351c9cf336c0e306a1c5d1/pkg/kv/kvserver/replica.go#L2936)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Instead of the nesting here, could we print two booleans in one log statements.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dodeca12 commented:
bors r+

dodeca12 commented:
bors r-

dodeca12 commented:
bors r+
